### PR TITLE
Update authenticate and authenticators API

### PR DESCRIPTION
**Pull request by DaazKu**
Partially address #6802

Check the specs: https://github.com/vanilla/roadmap/blob/master/api/

- Remove session related stuff from `/authenticate`
  - Disabled connect feature in AuthenticateController because of that.
- Add `/authenticate/authenticators` stuff
- Refactor POST `/authenticate` so that authenticators `type` and `id` are under the 'authenticate' field.
- Add a constraint on authenticator where type MUST equals id if the authenticator is unique. This helps getting authenticators by ID that that constraint will help "blacklisting" certain authenticator IDs in the UI.
- Add back getAuthenticatorByID in AuthenticatorModel. Always defaults to unique authenticator if the ID would clash.
- Add `isUserLinked()` function to SSOAuthenticators. This allows us to know if a user is linked or not to an authenticator using GET `/authenticate/authenticators` and `/authenticate/authenticators/:id`
- Add unit tests and update old ones.

### CODE REVIEW

File: library/Vanilla/Authenticator/PasswordAuthenticator.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ public function __construct(UserModel $userModel) {
     protected static function getAuthenticatorTypeInfoImpl(): array {
         return [
             'ui' => [
-                'photoUrl' => '/applications/dashboard/design/images/password-authenticator.svg',
+                'photoUrl' => '/applications/dashboard/design/images/password-Authenticator.svg',
```
Comment (initvector) [inline library/Vanilla/Authenticator/PasswordAuthenticator.php]: Probably wanna keep this lowercase

File: library/Vanilla/Models/AuthenticatorModel.php [hunk 2]
```diff
@@ -159,7 +184,24 @@ public function getAuthenticators() {
     }
 
     /**
-     * Get available authenticator classes.
+     * Get the list of ID of unique authenticators.
+     *
+     * @return array
+     */
+    public function getUniqueAuthenticatorIDs() {
```
Comment (initvector) [inline library/Vanilla/Models/AuthenticatorModel.php]: Is it safe to slap a return type on this method?

File: tests/APIv2/Authenticate/LinkUserTest.php [hunk 3]
```diff
@@ -145,15 +155,47 @@ public function testLinkUserWrongPassword() {
         $this->assertEquals($this->currentUser['userID'], $body['userID']);
     }
 
+    public function testUnlinkUser() {
```
Comment (initvector) [inline tests/APIv2/Authenticate/LinkUserTest.php]: Could use some doc-blockin'

File: applications/dashboard/controllers/api/AuthenticateApiController.php [hunk 4]
```diff
@@ -127,133 +130,128 @@ public function delete($authenticator, $authenticatorID = '', array $query) {
...
-     * @param array $query
-     * @return array
      * @throws Exception
+     * @param array $query
+     * @return Authenticator
      */
-    public function get_session($authSessionID, array $query) {
-        $this->permission();
+    public function index_authenticators(array $query) {
+         $this->permission();
```
Comment (initvector) [inline applications/dashboard/controllers/api/AuthenticateApiController.php]: One too many spaces, I think

File: applications/dashboard/controllers/class.authenticatecontroller.php [hunk 5]
```diff
@@ -104,26 +104,32 @@ public function initialize() {
...
         if ($response['authenticationStep'] === 'authenticated') {
             $redirectURL = (val('target', $this->request->getQuery(), '/'));
         } else {
-            $target = val('target', $this->request->getQuery());
-            if ($target) {
-                $target = '&target='.$target;
-            }
-            $redirectURL = "/authenticate/connectuser?authSessionID={$response['authSessionID']}{$target}";
+            throw new Gdn_ErrorException('Connect user feature is not implemented.');
+//            $target = val('target', $this->request->getQuery());
```
[the remark this answers is not in the document]
Comment (DaazKu) [reply applications/dashboard/controllers/class.authenticatecontroller.php]: Nah. I'm working in there next sprint and is really doesn't do much to have these completely removed vs commented.

**Pull request merged**